EcohydrologyTeam/ClearWater-riverine
# Add ability to model multiple constituents

Addresses #51 with the `constituents` module containing the `Constituent` class so that users can run multiple constituents at once. This will support integration with [ClearWater-modules](https://github.com/EcohydrologyTeam/ClearWater-modules) NSM. 

This update changes how the model is instantiated. We now set up the ClearwaterRiverine model with the `config_filepath` parameter. If a config is provided, the model will use all information from the config by default (i.e., it will override any of the other optional input parameters in the Python interface), therefore also addresses #48. See an example of a config:

```yaml
flow_field_filepath: W:\2ERDC12 - Clearwater\config\clearWaterTestCases.p48.hdf
diffusion_coefficient: 0.1
constituents:
  conservative_tracer:
    initial_conditions: W:\2ERDC12 - Clearwater\config\cwr_initial_conditions_tracer.csv
    boundary_conditions: W:\2ERDC12 - Clearwater\config\cwr_boundary_conditions_tracer.csv
    units: mg/m3
  temperature:
    initial_conditions: W:\2ERDC12 - Clearwater\config\cwr_initial_conditions_temp.csv
    boundary_conditions: W:\2ERDC12 - Clearwater\config\cwr_boundary_conditions_temp.csv
    units: degC
```
Users can still set up without a config, but would require the following parameters:
* `flow_field_file_path`
* `diffusion_coefficient`
* `constituent_dict`: a dictionary containing constituent names, initial conditions, boundary conditions, and optionally units (i.e., like the `constituents` section of the config dictionary). 

I pulled together a quick [demo notebook](https://github.com/EcohydrologyTeam/ClearWater-riverine/blob/constituents/examples/dev_sandbox/multiple_constituents.ipynb). 

Note that this PR is to merge to `constituents-dev` branch. I propose we keep this workflow there as we tie up the following loose ends, and then have a more comprehensive merge to main:
- Update `postproc_utils` so it will work with the new constituent setup (#76) 
  - Update pytests as well, if necessary 
- Update demo notebooks to use a config
- Update plotting functions to work with multiple constituents [and, more broadly, to work at all] (#77)

My goal with this PR is to get the initial implementation into the `constituents-dev` branch and make sure the general set-up and framework look good to you both. We can then split up the remaining tasks to continue merging to the `constituents-dev` branch until we feel that this change is polished enough for the main branch. Then I think we'll be ready for a release!
